# Проект по SQL  (финальный)  

### Цели исследования
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.    
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая **задача как аналитика — проанализировать базу данных**.    
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут **сформулировать ценностное предложение для нового продукта**.  

### Описание данных    

**Таблица books**  
Содержит данные о книгах:  
- book_id — идентификатор книги;  
- author_id — идентификатор автора;  
- title — название книги;  
- num_pages — количество страниц;  
- publication_date — дата публикации книги;  
- publisher_id — идентификатор издателя.  
**Таблица authors**  
Содержит данные об авторах:  
 - author_id — идентификатор автора;  
 - author — имя автора.  
**Таблица publishers**  
Содержит данные об издательствах:  
 - publisher_id — идентификатор издательства;  
 - publisher — название издательства;  
**Таблица ratings**  
Содержит данные о пользовательских оценках книг:  
 - rating_id — идентификатор оценки;  
 - book_id — идентификатор книги;  
 - username — имя пользователя, оставившего оценку;  
 - rating — оценка книги.  
**Таблица reviews**  
Содержит данные о пользовательских обзорах:  
 - review_id — идентификатор обзора;  
 - book_id — идентификатор книги;  
 - username — имя автора обзора;  
 - text — текст обзора.  






<a id='step_0'></a><h3>План работы</h3>


[1. Подключиться к БД.](#step_1)   
[2. Провести исследование таблиц — вывести первые строки.](#step_2)    
[3. Посчитать, сколько книг вышло после 1 января 2000 года.](#step_3)  
[4. Для каждой книги посчитать количество обзоров и среднюю оценку.](#step_4)  
[5. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц.](#step_5)   
[6. Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.](#step_6)   
[7. Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.](#step_7)   
[8. Выводы](#step_8)

<a id='step_1'></a><h3>1. Подключиться к БД</h3>

[К плану работ](#step_0)  


In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': ______, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas



<a id='step_2'></a><h3>2. Провести исследование таблиц — вывести первые строки.</h3>

[К плану работ](#step_0)  


In [2]:
query = '''SELECT * FROM books;'''
con=engine.connect()
df = pd.io.sql.read_sql(sql=text(query), con = con)
display(df.head())
df.info()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [3]:
query = '''
SELECT * 
FROM authors; 
'''
con=engine.connect()
df=pd.io.sql.read_sql(sql=text(query), con = con)
display(df.head())
df.info()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [4]:
query = '''
SELECT * 
FROM ratings; 
'''
con=engine.connect()
df=pd.io.sql.read_sql(sql=text(query), con = con)
display(df.head())
df.info()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [5]:
query = '''
SELECT * 
FROM reviews; 
'''
con=engine.connect()
df=pd.io.sql.read_sql(sql=text(query), con = con)
display(df.head())
df.info()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


In [6]:
query = '''
SELECT * 
FROM publishers 
'''
con=engine.connect()
df=pd.io.sql.read_sql(sql=text(query), con = con)
display(df.head())
df.info()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Верх таблиц выведен, есть строковые и численные форматы. Всего 1000 книг, 636 авторов, 340 издательств,  6456 рейтингов от ревьюверов. Для Ленинской библиотеки маловато, но для мобильного приложения с оплачиваемым контентом вполне подходит.


<a id='step_3'></a><h3>3. Посчитать, сколько книг вышло после 1 января 2000 года.</h3>

[К плану работ](#step_0)  


In [7]:
query = '''
SELECT 
    COUNT(publisher_id)
FROM 
    books 
WHERE 
    CAST(publication_date AS date) > '2000-01-01';
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

После 1 января 2000 года вышло 819 книг.

 
<a id='step_4'></a><h3>4. Для каждой книги посчитать количество обзоров и среднюю оценку.</h3>

[К плану работ](#step_0)  


In [8]:
##Для каждой книги посчитайте количество обзоров и среднюю оценку;
query = '''
SELECT a2.title AS title,
       a2.cnt_rev AS cnt_rev,
       AVG(a2.rating) AS avg_rat
FROM(
        (SELECT b.book_id AS book_id_,
                b.title AS title,
                COUNT(r.review_id) AS cnt_rev
         FROM books AS b
         LEFT JOIN reviews AS r ON r.book_id=b.book_id
         GROUP BY book_id_,
                  title
        ) AS a1
                  
     LEFT JOIN ratings AS r ON r.book_id=a1.book_id_) AS a2
GROUP BY
        title,
        cnt_rev
ORDER BY
         avg_rat DESC,
         cnt_rev DESC
LIMIT 5;
'''
con=engine.connect()
df = pd.io.sql.read_sql(sql=text(query), con = con)
df


title  cnt_rev  avg_rat
0                      A Dirty Job (Grim Reaper  #1)        4      5.0
1            School's Out—Forever (Maximum Ride  #2)        3      5.0
2       Moneyball: The Art of Winning an Unfair Game        3      5.0
3  Wherever You Go  There You Are: Mindfulness Me...        2      5.0
4               Pop Goes the Weasel (Alex Cross  #5)        2      5.0

Максимальный рейтинг 5, количество обзоров -4.  
В первую пятерку попадают:  
1. A Dirty Job (Grim Reaper #1)  
2. School's Out—Forever (Maximum Ride #2)  
3. Moneyball: The Art of Winning an Unfair Game  
4. Wherever You Go There You Are: Mindfulness Me  
5. Pop Goes the Weasel (Alex Cross #5)

 
<a id='step_5'></a><h3>5. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц.</h3>

[К плану работ](#step_0)  

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключим из анализа брошюры;  
 

In [9]:
##Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
query = '''
SELECT
    books.publisher_id,
    publishers.publisher,
    COUNT(books.publisher_id) AS cnt_book
FROM
    books
INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE
     books.num_pages>50
GROUP BY
        books.publisher_id,
        publishers.publisher
ORDER BY
        cnt_book DESC
LIMIT 1;

        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id      publisher  cnt_book
0           212  Penguin Books        42

Издательство Penguin Books самое плодотворное - оно выпустила 42 книги с количеством страниц более 50.


<a id='step_6'></a><h3>6. Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.</h3>

[К плану работ](#step_0)  


In [10]:
# Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

query = ''' SELECT
                a1.author,
                AVG (a1.avg_book_rat) AS avg_auth_rat
            FROM (
                    SELECT
                        b.author_id,
                        b.book_id,
                        a.author_id,
                        a.author,
                        r.book_id,
                        AVG (r.rating) AS avg_book_rat,
                        COUNT(r.rating) AS cnt_rat
                    FROM
                        books AS b
                        INNER JOIN authors AS a ON b.author_id = a.author_id
                        INNER JOIN ratings AS r ON b.book_id = r.book_id
                    GROUP BY
                        b.author_id,
                        b.book_id,
                        a.author_id,
                        a.author,
                        r.book_id
                    HAVING
                        COUNT(r.rating)>50
                    ORDER BY
                        b.author_id) AS a1
                GROUP BY
                    a1.author
                ORDER BY
                    avg_auth_rat DESC
                LIMIT 1    
        '''
df = pd.io.sql.read_sql(query, con=engine)
df

author  avg_auth_rat
0  J.K. Rowling/Mary GrandPré      4.283844

Джоан Роулинг имеет самый высокий рейтинг среди самых оцениваемых книг(более 50 оценок). Автор Гарри Поттера - не удивительно, ведь 2000 год она была на пике популярности.

<a id='step_7'></a><h3>7. Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</h3>

[К плану работ](#step_0)  

In [11]:
# Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок
query = '''
SELECT AVG(a1.cnt_rev) AS avg_cnt_rew
       
FROM(
     SELECT
           r.username,
           COUNT(r.review_id) AS cnt_rev
     FROM 
          reviews AS r
     WHERE username IN (
                        SELECT username
                        FROM ratings
                        GROUP BY username
                        HAVING COUNT(rating_id)>48
                        )
     GROUP BY username
     ORDER BY
     cnt_rev DESC) AS a1;
   
--LIMIT 10;
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)


avg_cnt_rew
0         24.0

Активные пользователи пишут обзоров в среднем в 2 раза меньше (24) чем ставят оценок (48)

 <a id='step_8'></a><h3>8. Выводы</h3>

[К плану работ](#step_0)  

База насчитывает 1000 книг, в ней есть популярные авторы такие как Джоан Роулинг, обзоры и оценки к книгам, издательства выпускают книги и брошюры. Есть солидные издательства с выпуском 43-х книг. Если правильно распорядиться этой информацией, то получится хороший стартап.